### ISSUE TITLE
Futuro algoritmo de mudança de peso

### ISSUE THREAD

**Issue by zito179** [zito179/Love-game#1]
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures
from datetime import datetime, timedelta

# ======================
# 1. DADOS COMPLETOS
# ======================
altura = 1.62
meta_peso_min, meta_peso_max = 68, 72

# Dados históricos
pesos = [58, 59, 65, 66, 68.5]  # kg
circ_abdominal = [68, 67, 66, 65, 64]  # cm

# Criar datas reais
data_inicio = datetime.now() - timedelta(days=len(pesos) - 1)
datas = [data_inicio + timedelta(days=i) for i in range(len(pesos))]

# ======================
# 2. PRÉ-PROCESSAMENTO
# ======================
dias = np.arange(len(datas)).reshape(-1, 1)

# Criar DataFrame
df = pd.DataFrame({
    'Data': datas,
    'Peso': pesos,
    'Circ_Abdominal': circ_abdominal
})

# Cálculo de variações de peso
df['Variação_Peso'] = df['Peso'].diff().fillna(0)

# Classificação da variação
df['Classificação'] = pd.cut(
    abs(df['Variação_Peso']),
    bins=[0, 1, 2, np.inf],
    labels=['Normal', 'Moderada', 'Extrema']
)

# ======================
# 3. MODELAGEM
# ======================
modelo_linear = LinearRegression().fit(dias, pesos)

poly = PolynomialFeatures(degree=2)
dias_poly = poly.fit_transform(dias)
modelo_poly = LinearRegression().fit(dias_poly, pesos)

# Suavização com média móvel
janela = 3
pesos_suavizados = np.convolve(pesos, np.ones(janela)/janela, mode='valid')

# ======================
# 4. PROJEÇÕES
# ======================
dias_futuros = np.arange(len(dias), len(dias) + 60).reshape(-1, 1)
dias_futuros_poly = poly.transform(dias_futuros)

proj_linear = modelo_linear.predict(dias_futuros)
proj_poly = modelo_poly.predict(dias_futuros_poly)

# Previsão de tempo para atingir metas
peso_atual = pesos[-1]
dias_ate_meta_min = (meta_peso_min - peso_atual) / modelo_linear.coef_[0]
dias_ate_meta_max = (meta_peso_max - peso_atual) / modelo_linear.coef_[0]

# ======================
# 5. ANÁLISE DE SAÚDE
# ======================
imc_atual = pesos[-1] / (altura ** 2)
taxa_semanal = modelo_linear.coef_[0] * 7

status_imc = (
    "Abaixo do peso" if imc_atual < 18.5 else
    "Normal" if 18.5 <= imc_atual < 25 else
    "Sobrepeso" if 25 <= imc_atual < 30 else "Obesidade"
)

classificacao_evolucao = (
    "Muito rápida" if taxa_semanal > 1 else
    "Saudável" if 0.2 <= taxa_semanal <= 1 else
    "Instável" if taxa_semanal < 0.2 else "Preocupante"
)

# ======================
# 6. VISUALIZAÇÃO
# ======================
fig, axs = plt.subplots(2, 1, figsize=(14, 10))

# Gráfico de peso
axs[0].plot(df['Data'], df['Peso'], 'o-', label='Peso Real', color='blue')
axs[0].plot(df['Data'], modelo_linear.predict(dias), 'g--', label='Tendência Linear')
axs[0].plot(df['Data'][janela-1:], pesos_suavizados, 'm:', label=f'Média Móvel ({janela} dias)')
datas_futuras = [df['Data'].iloc[-1] + timedelta(days=i) for i in range(1, 61)]
axs[0].plot(datas_futuras, proj_linear, 'r--', label='Projeção Linear (60 dias)')
axs[0].plot(datas_futuras, proj_poly, 'c--', label='Projeção Polinomial')
axs[0].fill_between(df['Data'], meta_peso_min, meta_peso_max, color='green', alpha=0.1, label='Faixa Ideal')
axs[0].set_title('Evolução do Peso')
axs[0].set_xlabel('Data')
axs[0].set_ylabel('Peso (kg)')
axs[0].grid(True, linestyle='--', alpha=0.7)
axs[0].legend()

# Gráfico de circunferência abdominal
axs[1].plot(df['Data'], df['Circ_Abdominal'], 'o-', label='Circunferência Abdominal', color='purple')
axs[1].set_title('Evolução da Circunferência Abdominal')
axs[1].set_xlabel('Data')
axs[1].set_ylabel('Circunferência (cm)')
axs[1].grid(True, linestyle='--', alpha=0.7)
axs[1].legend()

plt.tight_layout()

# ======================
# 7. RELATÓRIO CONSOLIDADO
# ======================
print("\n" + "="*50)
print(" RELATÓRIO DE ANÁLISE DE PESO ".center(50, '='))
print("="*50)

print(f"\n📊 DADOS ATUAIS:")
print(f"- Último peso: {peso_atual:.1f} kg ({df['Data'].iloc[-1].strftime('%d/%m/%Y')})")
print(f"- IMC atual: {imc_atual:.1f} ({status_imc})")
print(f"- Circ. abdominal: {circ_abdominal[-1]} cm")

print(f"\n📈 TENDÊNCIA:")
print(f"- Variação diária: {modelo_linear.coef_[0]:.2f} kg/dia")
print(f"- Variação semanal: {taxa_semanal:.2f} kg/semana")
print(f"- Classificação da evolução: {classificacao_evolucao}")

print(f"\n🔍 PROJEÇÕES:")
print(f"- Linear (60 dias): {proj_linear[-1]:.1f} kg")
print(f"- Polinomial (60 dias): {proj_poly[-1]:.1f} kg")

if dias_ate_meta_min > 0:
    print(f"- Tempo estimado para atingir {meta_peso_min} kg: {dias_ate_meta_min:.0f} dias")
if dias_ate_meta_max > 0:
    print(f"- Tempo estimado para atingir {meta_peso_max} kg: {dias_ate_meta_max:.0f} dias")

print(f"\n⚠️ ALERTAS:")
if any(df['Classificação'] == 'Extrema'):
    print("- 🚨 Variações extremas detectadas!")
if taxa_semanal > 0.5:
    print("- ⚠️ Ganho de peso acima do recomendado (OMS: até 0.5kg/semana)")

print("\n📋 HISTÓRICO COMPLETO:")
print(df[['Data', 'Peso', 'Variação_Peso', 'Classificação']].to_string(index=False))

plt.show()

Oque acham deste código
